---
title: "HIV CRISPR design"
author: "Pavitra Roychoudhury"
date: "January 27, 2015"
output: html_document
editor_options: 
  chunk_output_type: console
---

This script imports an HIV sequence in fasta format (typically a single subtype or group level consensus sequence for a specified gene from the LANL HIV sequence database) and first identifies all occurrences of a specified PAM sequence (NNGRRT for saCas9 and NGG for spCas9, R = A/G, N = A/C/T/G). For each PAM sequence, it returns the sequence 20 bases upstream (i.e. target sequence) and counts the number of exact matches within the input sequence. Next we import fasta files containing 1) subtype-level consensus sequences and 2) some larger list of representative sequences. We then tabulate how many times each pam+target occurs within each of these lists. 

Analyses for Figures 1 & 2, Table 1, S1 and S2


### 1. Update packages and load libraries
```{r,message=F}
rm(list=ls()); 
sessionInfo();
options(error=recover);
# source("http://bioconductor.org/biocLite.R");
# biocLite(ask=FALSE); # run biocLite("BiocUpgrade") if upgrade is required

# library('ShortRead'); 
library('Biostrings'); library(Rsamtools)
library('stringr');library('dplyr')
```

### 2. Set directories, import data
This is where you specify the PAM sequence, filename of the reference sequence and where to save any output files generated by this script. Un-comment one chunk at a time to run the script for one gene


LTR:
```{r}
# output_dir<-'./results_SpCas9_ltr/';
# 
# #allseqsfname<-'./alignments/HIV1_ALL_2014_ltr_DNA'; #2014 one was used for the original design.
# allseqsfname<-'./alignments/HIV1_ALL_2016_ltr_DNA';
# 
# refseqdir<-'./consensus_seqs/ltr/';
# 
# sgrnadesigner_folder<-'./sgRNA_designer_broad/ltr/';
# 
# grps<-c('grpM','subA','subB','subC'); #groups/subtypes we want to look at

```


GAG:
```{r}
# output_dir<-'./results_SpCas9_gag/';
# 
# allseqsfname<-'./alignments/HIV1_ALL_2016_gag_DNA'; #bam file that contains db of seqs, ignore extension
# 
# refseqdir<-'./consensus_seqs/gag/';
# 
# sgrnadesigner_folder<-'./sgRNA_designer_broad/gag/';
# 
# grps<-c('grpM','subA','subB','subC'); #groups/subtypes we want to look at

```

Pol:
```{r}
output_dir<-'./results_SpCas9_pol/';

allseqsfname<-'./alignments/HIV1_ALL_2016_pol_DNA'; #bam file that contains db of seqs, ignore extension

refseqdir<-'./consensus_seqs/pol/';

sgrnadesigner_folder<-'./sgRNA_designer_broad/pol/';

grps<-c('grpM','subA','subB','subC'); #groups/subtypes we want to look at
```


Common settings: 
We are running this for spCas9, if saCas9, change pam to match: 
pattern_fwd<-DNAString('NNNRRT');
```{r}
pattern_fwd<-DNAString('NGG');

guide_length<-20; #not including the pam sequence
```


Import required files:
```{r}
#Import bam file containing sequences from the database
params<-ScanBamParam(flag=scanBamFlag(isUnmappedQuery=FALSE),
                     what=c('qname','strand','qwidth','cigar','seq'));
reads<-scanBam(file=paste(allseqsfname,'.bam',sep=''),
               index=paste(allseqsfname,'.bai',sep=''),
               param=params); #import all elements from bam file
rm(params);
names(reads[[1]]); #this lists all the elements in the bam file
head(reads[[1]]$qname); #read IDs
head(reads[[1]]$cigar); #the alignment
reads[[1]]$seq; #sequences in the bam file

#Make a table of sequences and subtypes
subtypes<-unlist(lapply(reads[[1]]$qname,function(x)strsplit(x,'\\.')[[1]][1]));
aligned_seqs_table<-data.frame(seqname=reads[[1]]$qname,
                               width=reads[[1]]$qwidth,
                               subtype=subtypes);
head(aligned_seqs_table);

write.csv(aligned_seqs_table,
          file=paste(output_dir,'Aligned_seqs_info_',basename(allseqsfname),'.csv',sep=''),row.names=F);

aligned_seq_counts<-aligned_seqs_table%>%group_by(subtype)%>%
  summarize(tot_seqs=n());
write.csv(aligned_seq_counts,
  file=paste(output_dir,'Aligned_seqs_subtypecounts_',basename(allseqsfname),'.csv',sep=''),
  row.names=F)
```



### 3. Find candidate guides and calculate targetted seqs and efficacy scores

Adds scores from sgRNA designer: 
Previously we were using Benchling + Offspotter. But the current version of sgRNA designer from the Broad Institute incorporates both on-target scores as well as off-target. The on-target score is based on Rule Set 2 of Doench et al Nat Biotech 2016. The on-target hits are scored in Tiers I-IV based on whether the hits are in coding regions, non-coding regions of coding genes, non-coding regions of non-coding genes, or all other regions. The scores are binned into 4 bins based on the CFD (cutting frequency determination) score. 

```{r}
for(grp in grps){
  print(grp);
  
  #Import reference fasta file
  refseqname<-grep(grp,list.files(refseqdir,full.names=T),value=T);
  ref_seq<-DNAString(as.character(sread(readFasta(refseqname))));
  ref_seq;
  
  #Create reverse complement
  ref_seq_rev<-reverseComplement(ref_seq);
  ref_seq_rev;
  
  #Look for PAM sequence matches: Search reference sequence for PAM sequence 
  # and return hits in forward and reverse direction.
  matches_fwd<-matchPattern(pattern_fwd,ref_seq,fixed=F);
  matches_rev<-matchPattern(pattern_fwd,ref_seq_rev,fixed=F);
  matches_fwd;
  matches_rev;
  print(length(matches_fwd)+length(matches_rev))
  
  #Find candidate targets: Get target seqs (fwd and rev) for each PAM sequence match from above.
  # Write the results to a fasta file in case you want to do alignments/analyses elsewhere.

  #Forward
  upstream_inds_fwd<-IRanges(start(matches_fwd)-guide_length,
                             width=guide_length+nchar(pattern_fwd));
  upstream_inds_fwd<-upstream_inds_fwd[start(upstream_inds_fwd)>0];
  guide_seqs_fwd<-sapply(upstream_inds_fwd,function(x)return(as.character(ref_seq[x])));
  rm(upstream_inds_fwd);
  # tmpdf<-data.frame(label=paste('>seq',seq(1,length(guide_seqs_fwd)),sep=''),
  #                   seq=guide_seqs_fwd);
  # write.table(c(t(tmpdf)),file=paste(output_dir,'All_guide_seqs_fwd.fa',sep=''),
  #             sep='\r\n',row.names=F,col.names=F,quote=F);
  
  #Reverse
  upstream_inds_rev<-IRanges(start(matches_rev)-guide_length,
                             width=guide_length+nchar(pattern_fwd));
  upstream_inds_rev<-upstream_inds_rev[start(upstream_inds_rev)>0];
  guide_seqs_rev<-sapply(upstream_inds_rev,function(x)return(as.character(ref_seq_rev[x])));
  rm(upstream_inds_rev);
  # tmpdf<-data.frame(label=paste('>seq',seq(1,length(guide_seqs_rev)),sep=''),
  #                   seq=guide_seqs_rev);
  # write.table(c(t(tmpdf)),file=paste(output_dir,'All_guide_seqs_rev.fa',sep=''),
  #             sep='\r\n',row.names=F,col.names=F,quote=F);
  
  #Preview candidate guides (with the pam sequence)
  head(guide_seqs_fwd,10);
  head(guide_seqs_rev,10);
  
  #Remove duplicates
  guide_seqs_fwd<-unique(guide_seqs_fwd);
  guide_seqs_rev<-unique(guide_seqs_rev);
  print(length(guide_seqs_fwd)+length(guide_seqs_rev));
  
  #Remove any guides with ambiguous characters
  if(any(grepl('M|R|W|S|Y|K|V|H|D|B|N|-',guide_seqs_fwd))){
  	guide_seqs_fwd<-guide_seqs_fwd[-grep('M|R|W|S|Y|K|V|H|D|B|N|-',guide_seqs_fwd)];
  	guide_seqs_rev<-guide_seqs_rev[-grep('M|R|W|S|Y|K|V|H|D|B|N|-',guide_seqs_rev)];
  }
  print(length(guide_seqs_fwd)+length(guide_seqs_rev));
  
  #Analyze guide frequency: For each candidate guide, record direction it was found, length
  # (they're all the same length),and whether there's a G in position 1, 2 and 3. 
  # Then look for the number of occurrences of each guide in the forward and reverse 
  # directions within the list of sequences for all group M.
  
  #Analyze guides
  candidate_guides<-data.frame(
    gseq=c(guide_seqs_fwd,guide_seqs_rev),
    dir=c(rep('F',length(guide_seqs_fwd)),
          rep('R',length(guide_seqs_rev))),
    from_consensus=grp,stringsAsFactors=F)%>%
    mutate(length=str_length(gseq),
           guide=substr(gseq,1,guide_length),
           pam=substr(gseq,guide_length+1,guide_length+nchar(pattern_fwd)),
           g_1=str_sub(gseq,1,1)=='G',
           g_2=str_sub(gseq,2,2)=='G',
           g_3=str_sub(gseq,3,3)=='G');
  head(candidate_guides);
  
  #Search for guides in alignment -- exact matches
  candidate_guides$gpM_freq_fwd<-vcountPDict(PDict(candidate_guides$gseq),reads[[1]]$seq,
                                             max.mismatch=0,collapse=1);
  candidate_guides$gpM_freq_rev<-vcountPDict(PDict(candidate_guides$gseq),
                                             reverseComplement(reads[[1]]$seq),
                                             max.mismatch=0,collapse=1);
  candidate_guides$subA_freq_fwd<-vcountPDict(PDict(candidate_guides$gseq),
                                              reads[[1]]$seq[subtypes%in%c('A','A1','A2','A3','A4','A5','A6')],
                                              max.mismatch=0,collapse=1);
  candidate_guides$subA_freq_rev<-vcountPDict(PDict(candidate_guides$gseq),
                                              reverseComplement(reads[[1]]$seq[subtypes%in%c('A','A1','A2','A3','A4','A5','A6')]),
                                              max.mismatch=0,collapse=1);
  candidate_guides$subB_freq_fwd<-vcountPDict(PDict(candidate_guides$gseq),
                                              reads[[1]]$seq[subtypes=='B'],
                                              max.mismatch=0,collapse=1);
  candidate_guides$subB_freq_rev<-vcountPDict(PDict(candidate_guides$gseq),
                                              reverseComplement(reads[[1]]$seq[subtypes=='B']),
                                              max.mismatch=0,collapse=1);
  candidate_guides$subC_freq_fwd<-vcountPDict(PDict(candidate_guides$gseq),
                                              reads[[1]]$seq[subtypes=='C'],
                                              max.mismatch=0,collapse=1);
  candidate_guides$subC_freq_rev<-vcountPDict(PDict(candidate_guides$gseq),
                                              reverseComplement(reads[[1]]$seq[subtypes=='C']),
                                              max.mismatch=0,collapse=1);
  candidate_guides$other_freq_fwd<-vcountPDict(PDict(candidate_guides$gseq),
                                              reads[[1]]$seq[!subtypes%in%c('A','A1','A2','A3','A4','A5','A6','B','C')],
                                              max.mismatch=0,collapse=1);
  candidate_guides$other_freq_rev<-vcountPDict(PDict(candidate_guides$gseq),
                                              reverseComplement(reads[[1]]$seq[!subtypes%in%c('A','A1','A2','A3','A4','A5','A6','B','C')]),
                                              max.mismatch=0,collapse=1);
  
  candidate_guides<-mutate(candidate_guides,
                           gpM_freq_total=gpM_freq_fwd+gpM_freq_rev,
                           subA_freq_total=subA_freq_fwd+subA_freq_rev,
                           subB_freq_total=subB_freq_fwd+subB_freq_rev,
                           subC_freq_total=subC_freq_fwd+subC_freq_rev,
                           other_freq_total=other_freq_fwd+other_freq_rev,
                           gpM_perc=100*gpM_freq_total/nrow(aligned_seqs_table),
                           subA_perc=100*subA_freq_total/sum(aligned_seqs_table$subtype%in%c('A','A1','A2','A3','A4','A5','A6')),
                           subB_perc=100*subB_freq_total/sum(aligned_seqs_table$subtype=='B'),
                           subC_perc=100*subC_freq_total/sum(aligned_seqs_table$subtype=='C'),
                           other_perc=100*other_freq_total/sum(!aligned_seqs_table$subtype%in%c('A','A1','A2','A3','A4','A5','A6','B','C')));
  head(candidate_guides);
  
  #Add scores from sgRNA designer
  guidescores<-read.delim(grep(grp,list.files(sgrnadesigner_folder,full.names=T),value=T),stringsAsFactors=F);
  
  ##Leaving out off target scores for now and only using the ranking. 
  ##If required later on, need to use some version of the following:
  # guidescores[,grepl('Off.Target.Tier.*.Matches',names(guidescores))]
  # guidescores$total_offtgt<-apply(guidescores[,grepl('Off.Target.Tier.*.Matches',names(guidescores))],1,sum)
  # candidate_guides[,c('broad_rank','efficacy_score')]<-
  # 	guidescores[unlist(lapply(candidate_guides$guide,function(x)
  # 		match(x,guidescores$sgRNA.Sequence))),c('Combined.Rank','On.Target.Efficacy.Score')];
  
  candidate_guides<-cbind(candidate_guides,
  	guidescores[unlist(lapply(candidate_guides$guide,function(x)
  		match(x,guidescores$sgRNA.Sequence))),]);
  head(candidate_guides);
  
  write.csv(candidate_guides,file=paste(output_dir,'Single_guides_',grp,'.csv',sep=''),row.names=F);
}
```


### 4. Pool guides and get matches to alignment
Pool guides from all subtypes, clean up, sort by frequency
Only needs steps 1 and 2 to be run before this.
```{r}
#Pool all guides we have so far
all_guides_df<-do.call('rbind',lapply(grps,function(x){
  read.csv(paste(output_dir,'Single_guides_',x,'.csv',sep=''),stringsAsFactors=F)
}));
nrow(all_guides_df);
head(all_guides_df)

#Remove duplicates
all_guides_df<-distinct(all_guides_df,gseq,.keep_all=T)
nrow(all_guides_df);

#Remove guides with no hits, sort by frequency
all_guides_df<-filter(all_guides_df,gpM_freq_total>0);
all_guides_df<-all_guides_df[order(all_guides_df$gpM_freq_total,decreasing=T),]%>%
  mutate(gid=c(1:nrow(all_guides_df)));
nrow(all_guides_df);
write.csv(all_guides_df,paste(output_dir,'All_guides.csv',sep=''),row.names=F)

all_guides<-all_guides_df$gseq;

#Matrix to store matches and visualize in a heatmap
#To do: replace the A-other bits with just indexing rather than repeating the vcount
match_matrix_gpM<-t(vcountPDict(PDict(all_guides),reads[[1]]$seq,
            max.mismatch=0,collapse=F)+
              vcountPDict(PDict(all_guides),reverseComplement(reads[[1]]$seq),
            max.mismatch=0,collapse=F));
match_matrix_subA<-t(vcountPDict(PDict(all_guides),
                                 reads[[1]]$seq[aligned_seqs_table$subtype%in%c('A','A1','A2','A3','A4','A5','A6')],
            max.mismatch=0,collapse=F)+vcountPDict(PDict(all_guides),
                          reverseComplement(reads[[1]]$seq[aligned_seqs_table$subtype%in%c('A','A1','A2','A3','A4','A5','A6')]),
            max.mismatch=0,collapse=F));
match_matrix_subB<-t(vcountPDict(PDict(all_guides),
                                 reads[[1]]$seq[aligned_seqs_table$subtype=='B'],
            max.mismatch=0,collapse=F)+vcountPDict(PDict(all_guides),
                          reverseComplement(reads[[1]]$seq[aligned_seqs_table$subtype=='B']),
            max.mismatch=0,collapse=F));
match_matrix_subC<-t(vcountPDict(PDict(all_guides),
                                 reads[[1]]$seq[aligned_seqs_table$subtype=='C'],
            max.mismatch=0,collapse=F)+vcountPDict(PDict(all_guides),
                          reverseComplement(reads[[1]]$seq[aligned_seqs_table$subtype=='C']),
            max.mismatch=0,collapse=F));
match_matrix_other<-t(vcountPDict(PDict(all_guides),
                                 reads[[1]]$seq[!aligned_seqs_table$subtype%in%c('A','A1','A2','A3','A4','A5','A6','B','C')],
            max.mismatch=0,collapse=F)+vcountPDict(PDict(all_guides),
                          reverseComplement(reads[[1]]$seq[!aligned_seqs_table$subtype%in%c('A','A1','A2','A3','A4','A5','A6','B','C')]),
            max.mismatch=0,collapse=F));
dim(match_matrix_gpM)
```





### 5. Make plots
```{r}
# match_matrix<-match_matrix_gpM;
# suffix<-'all';
#   
# #Max possible coverage
# cov_seq<-apply(match_matrix,1,function(x)sum(x)>0); #whether each seq is targetted by at least one guide
# print(sum(cov_seq))
# tot_seq<-length(reads[[1]]$seq);
# print(100*sum(cov_seq)/tot_seq)
# which(cov_seq==FALSE); #which seqs are not targetted
# aligned_seqs_table[which(cov_seq==FALSE),]
# 
# #Make heatmaps of all guides
# pdf(file=paste(output_dir,'Heatmap_guides_',suffix,'.pdf',sep=''),width=10,height=5)
# # svg(file=paste(output_dir,'Heatmap_guides_',suffix,'.svg',sep=''),width=10,height=5)
# par(mar=c(3,3,0.5,0.5))
# image(x=c(1:nrow(match_matrix)),y=c(1:ncol(match_matrix)),
#       z=match_matrix,col=c('white','#7fbf7b','red'),
#       xlab='',ylab='');
# rect(xleft=min(which(aligned_seqs_table$subtype%in%c('A','A1','A2','A3','A4','A5','A6'))),
#      ybottom=1,xright=max(which(aligned_seqs_table$subtype%in%c('A','A1','A2','A3','A4','A5','A6'))),
#      ytop=ncol(match_matrix),col=rgb(255/255,182/255,193/255,0.3),border=NA)
# rect(xleft=min(which(aligned_seqs_table$subtype=='B')),
#      ybottom=1,xright=max(which(aligned_seqs_table$subtype=='B')),
#      ytop=ncol(match_matrix),col=rgb(135/255,206/255,250/255,0.3),border=NA);
# rect(xleft=min(which(aligned_seqs_table$subtype=='C')),
#      ybottom=1,xright=max(which(aligned_seqs_table$subtype=='C')),
#      ytop=ncol(match_matrix),col=rgb(255/255,255/255,0/255,0.3),border=NA);
# legend('topright',legend=c('A','B','C'),fill=c(rgb(255/255,182/255,193/255,0.3),
#                                                rgb(135/255,206/255,250/255,0.3),
#                                                rgb(255/255,255/255,0/255,0.3)),
#        bty='n',title='Subtype');
# mtext('LANL sequence',side=1,line=2);
# mtext('Guide #',side=2,line=2);
# dev.off()


#For Harshana, 14-Aug-17
# harsh_list<-read.csv('./flow_guideactivity/harsh_list.csv',stringsAsFactors=F)
# match_matrix<-t(vcountPDict(PDict(paste(harsh_list$sequence,harsh_list$PAM,sep='')),reads[[1]]$seq,
#             max.mismatch=0,collapse=F)+
#               vcountPDict(PDict(paste(harsh_list$sequence,harsh_list$PAM,sep='')),reverseComplement(reads[[1]]$seq),
#             max.mismatch=0,collapse=F));
# 
# pdf(file=paste(output_dir,'Heatmap_guides_Harsh.pdf',sep=''),width=10,height=3)
# par(mar=c(3,3,0.5,0.5))
# image(x=c(1:nrow(match_matrix)),y=c(1:ncol(match_matrix)),
#       z=match_matrix,col=c('white','#7fbf7b'),
#       xlab='',ylab='');
# rect(xleft=min(which(aligned_seqs_table$subtype%in%c('A','A1','A2','A3','A4','A5','A6'))),
#      ybottom=1,xright=max(which(aligned_seqs_table$subtype%in%c('A','A1','A2','A3','A4','A5','A6'))),
#      ytop=ncol(match_matrix),col=rgb(255/255,182/255,193/255,0.3),border=NA)
# rect(xleft=min(which(aligned_seqs_table$subtype=='B')),
#      ybottom=1,xright=max(which(aligned_seqs_table$subtype=='B')),
#      ytop=ncol(match_matrix),col=rgb(135/255,206/255,250/255,0.3),border=NA);
# rect(xleft=min(which(aligned_seqs_table$subtype=='C')),
#      ybottom=1,xright=max(which(aligned_seqs_table$subtype=='C')),
#      ytop=ncol(match_matrix),col=rgb(255/255,255/255,0/255,0.3),border=NA);
# legend('topright',legend=c('A','B','C'),fill=c(rgb(255/255,182/255,193/255,0.3),
#                                                rgb(135/255,206/255,250/255,0.3),
#                                                rgb(255/255,255/255,0/255,0.3)),
#        bty='n',title='Subtype');
# mtext('LANL sequence',side=1,line=2);
# mtext('Guide #',side=2,line=2);
# dev.off()

#Total hits (ranked)
# library(ggplot2);
# pdf(file=paste(output_dir,'Total_hits.pdf',sep=''),width=5,height=3);
# ggplot(all_guides_df,aes(x=gid,y=gpM_perc))+
#   geom_bar(stat='identity')+
#   xlab('Guide #')+ylab('Total grp M hits')+ylim(0,100)+
#   theme_bw()
# ggplot(all_guides_df,aes(x=gid,y=subA_perc))+
#   geom_bar(stat='identity',fill=rgb(255/255,182/255,193/255))+
#   xlab('Guide #')+ylab('Total sub A hits')+ylim(0,100)+
#   theme_bw();
# ggplot(all_guides_df,aes(x=gid,y=subB_perc))+
#   geom_bar(stat='identity',fill=rgb(135/255,206/255,250/255))+
#   xlab('Guide #')+ylab('Total sub B hits')+ylim(0,100)+
#   theme_bw();
# ggplot(all_guides_df,aes(x=gid,y=subC_perc))+
#   geom_bar(stat='identity',fill=rgb(255/255,255/255,0/255))+
#   xlab('Guide #')+ylab('Total sub C hits')+ylim(0,100)+
#   theme_bw();
# dev.off();
```

For Figure 1: Top 100 guides and number of sequences in each alignment
```{r}
library(reshape2);
tot_seq<-length(reads[[1]]$seq);
barplot_df<-all_guides_df[1:100,]%>%
  mutate(a_perc=subA_freq_total/tot_seq*100,
         b_perc=subB_freq_total/tot_seq*100,
         c_perc=subC_freq_total/tot_seq*100,
         other_perc=other_freq_total/tot_seq*100)%>%
  select(gid,a_perc,b_perc,c_perc,other_perc)%>%
  melt(id='gid');

# pdf(file=paste(output_dir,'Top100_barplot.pdf',sep=''),width=9,height=4);
svg(file=paste(output_dir,'Top100_barplot.svg',sep=''),width=7,height=3);
ggplot(barplot_df,aes(x=gid,y=value))+
  geom_bar(stat='identity',aes(fill=variable))+coord_flip()+
  scale_fill_brewer(palette='Accent',
                    name='Subtype',
                    labels=c('A','B','C','Other'))+
  ylab('Hits in Group M (%)')+xlab('Guide (ranked by total hits)')+
	ylim(0,100)+
  theme_bw();
dev.off();


#Plot number of sequences in the alignment (Figure 1, inset boxes in each panel)
require(reshape2);require(ggplot2)
head(aligned_seqs_table);
aligned_seq_summary<-aligned_seqs_table%>%
	summarize(A=sum(subtype%in%c('A','A1','A2','A3','A4','A5','A6')),
						B=sum(subtype=='B'),
						C=sum(subtype=='C'),
						Other=nrow(aligned_seqs_table)-(A+B+C))%>%
	melt();
nrow(aligned_seqs_table)

# pdf(file=paste(output_dir,'Seq_counts.pdf',sep=''),width=3,height=3);
svg(file=paste(output_dir,'Seq_counts.svg',sep=''),width=2,height=2);
ggplot(aligned_seq_summary,aes(x=variable,y=value))+
	# geom_col(fill='blue')+ #LTR
	# geom_col(fill='pink')+ #gag
	# geom_col(fill='red')+ #pol
	geom_col(aes(fill=variable))+
	scale_fill_brewer(palette='Accent',
                    name='Subtype',
                    labels=c('A','B','C','Other'))+
	xlab('Subtype')+ylab('# sequences')+
	theme_bw()+
	theme(legend.position='none');
dev.off()
```

Guide rank from sgRNA designer vs. diversity rank
```{r}
# library(ggplot2)
# # all_guides<-subset(read.csv('./results_SpCas9_ltr/AllGuides.csv',stringsAsFactors=F),from_consensus=='grpM')
# head(all_guides)
# all_guides$div_rank_M<-order(all_guides$gpM_perc,decreasing=T);
# # ggplot(all_guides,aes(x=div_rank_M,y=broad_rank))+
# # 	geom_point()+theme_bw()
# # ggplot(all_guides,aes(x=div_rank_M,y=efficacy_score))+
# # 	geom_point()+theme_bw()
# pdf('./results_SpCas9_ltr/Efficacy_score_hist_groupM.pdf',width=3,height=3)
# ggplot(all_guides,aes(x=efficacy_score))+
# 	geom_histogram(binwidth=0.05,fill='black')+theme_bw()+
# 	xlab('sgRNA efficacy score')+ylab('Count')
# dev.off()
```

Histogram of predicted efficacy
This chunk can be run after 1&2 alone. This one computes the histogram for one gene at a time
Next chunk has this figure combined for all guides across genes
```{r}
# library(ggplot2)
# all_guides_df<-read.csv(paste(output_dir,'All_guides.csv',sep=''),stringsAsFactors=F);
# svg(file=paste(output_dir,'Pred_ontgt_eff_allguides.svg',sep=''),width=2,height=2);
# ggplot(subset(all_guides_df,!is.na(On.Target.Efficacy.Score)),
#        aes(x=On.Target.Efficacy.Score))+
# 	# geom_histogram(binwidth=0.02,fill='blue')+
#   # geom_histogram(binwidth=0.02,fill='pink')+
#   geom_histogram(binwidth=0.02,fill='red')+
#   theme_bw()+xlim(0,0.9)+
# 	xlab('sgRNA efficacy score')+ylab('Count')
# dev.off()
# print('Number of guides with scores:')
# nrow(subset(all_guides_df,!is.na(On.Target.Efficacy.Score)))
```

Figure 2: Conservation vs predicted efficacy
This chunk can be run independently. No previous chunks required.
```{r}
library(ggplot2); library(parallel);
output_dirs<-c('./results_SpCas9_ltr','./results_SpCas9_gag','./results_SpCas9_pol');
all_guides_df<-do.call('rbind',mclapply(output_dirs,function(x)
	cbind(read.csv(paste(x,'/All_guides.csv',sep=''),stringsAsFactors=F),
				data.frame(folder=basename(x),gene=strsplit(basename(x),'_')[[1]][3]))));
all_guides_df$excluded_window<-grepl('Outside Target Window|Transcriptional Incompatibility|Cloning Incompatibility',
																	all_guides_df$Picking.Notes)
all_guides_df$excluded_offtgt<-grepl('Off-target Match Bin I matches > 3',
																	all_guides_df$Picking.Notes);
head(all_guides_df)

#Alternative version but it's messy
#Classify each guide according to hits: Grp M, Sub A-C, Other, Excluded
# all_guides_df$classification<-apply(all_guides_df,1,function(x)
# 	c('M','A','B','C','Other')[which.max(as.numeric(x[c('gpM_perc','subA_perc','subB_perc','subC_perc','other_perc')]))]);
# # all_guides_df$classification[all_guides_df$excluded_offtgt]<-'Excl-OffTgt';
# svg('./All_guides_ontgt_vs_conservation_subtypes_combined.svg',width=5,height=2)
# ggplot(data=subset(all_guides_df,!is.na(On.Target.Efficacy.Score)&!excluded_offtgt),
# 			 aes(x=On.Target.Efficacy.Score,y=gpM_perc))+
# 	geom_point(aes(colour=classification))+
# 	scale_color_brewer(type='qual',palette=1)+
# 	ylim(0,100)+xlim(0,0.9)+
# 	xlab('Predicted activity score')+
# 	ylab('Hits in subtypes (%)')+
# 	# geom_point(data=subset(all_guides_df,!is.na(On.Target.Efficacy.Score)&excluded_offtgt),colour='red',alpha=0.5)+
# 	geom_hline(yintercept=50,linetype='dashed',colour='blue',alpha=0.5)+
# 	geom_vline(xintercept=0.2,linetype='dashed',colour='blue',alpha=0.5)+
# 	theme_bw()+
# 	facet_grid(classification~gene);
# dev.off()

require(reshape2)
all_guides_melted<-melt(all_guides_df[,c('gseq','from_consensus','gene','folder',
                                        'excluded_offtgt','On.Target.Efficacy.Score',
                                        'gpM_perc','subA_perc','subB_perc','subC_perc'),],
                        id.vars=c('gseq','from_consensus','gene','folder',
                                        'excluded_offtgt','On.Target.Efficacy.Score'),
                        value.name='hits',variable.name='subtype');
head(all_guides_melted)
all_guides_melted$subtype<-factor(all_guides_melted$subtype,
                                   levels=c('gpM_perc','subA_perc','subB_perc','subC_perc'),
                                   labels=c('M','A','B','C'));
head(all_guides_melted);

#Correlation between predicted activity and conservation
require(broom); require(dplyr); require(tidyr)
activity_corr<-all_guides_melted%>%
	filter(!excluded_offtgt)%>%
	group_by(subtype,gene)%>%
	do(tidy(cor.test(.$On.Target.Efficacy.Score,.$hits)))
write.csv(activity_corr,
          './results_SpCas9/Correlation_activity_conservation_grouped.csv',row.names=F)
activity_corr<-subset(all_guides_melted,subtype=='M')%>%
	filter(!excluded_offtgt)%>%
	group_by(gene)%>%
	do(tidy(cor.test(.$On.Target.Efficacy.Score,.$hits)))
write.csv(activity_corr,'./results_SpCas9/Correlation_activity_conservation_all.csv',row.names=F)

activity_summ<-group_by(all_guides_df,gene)%>%
  summarise(mean_act=mean(On.Target.Efficacy.Score,na.rm=T),
            sd_act=sd(On.Target.Efficacy.Score,na.rm=T),
            n_guides=n(),
            n_offtgt=sum(excluded_offtgt==TRUE))
write.csv(activity_summ,'./results_SpCas9/Activity_score_summary.csv',row.names=F)


#Want to count the number of guides that meet the following criteria: on tgt >0.2, hits>50%, no off tgt match bin Is 
library(dplyr);
summary_table<-group_by(all_guides_melted,subtype,gene)%>%
  summarise(npicked=sum(!is.na(On.Target.Efficacy.Score)&
	                  !excluded_offtgt&On.Target.Efficacy.Score>=0.2&hits>50));
summary_table$On.Target.Efficacy.Score<-0.8;#for position
summary_table$hits<-90;
write.csv(summary_table,'./Guidepicks_cons_activity.csv');

svg('./sgRNA_designer_broad/All_guides_ontgt_vs_conservation_combined.svg',width=5.2,height=6.5)
ggplot(data=subset(all_guides_melted,!is.na(On.Target.Efficacy.Score)&!excluded_offtgt),
			 aes(x=On.Target.Efficacy.Score,y=hits))+
	geom_point(colour='grey28',alpha=0.5)+
	ylim(0,100)+xlim(0,0.9)+
	xlab('Predicted activity score')+
	ylab('Hits in group or subtype (%)')+
  geom_point(data=subset(all_guides_melted,!is.na(On.Target.Efficacy.Score)&excluded_offtgt),
	           colour='red',alpha=0.5,shape=17)+
	geom_hline(yintercept=50,linetype='dashed',colour='blue',alpha=0.5)+
	geom_vline(xintercept=0.2,linetype='dashed',colour='blue',alpha=0.5)+
	theme_bw()+
	facet_grid(subtype~gene)+
  geom_text(data=summary_table,aes(label=npicked),colour='blue')
dev.off()



#Histograms with facets
svg(file=paste('./sgRNA_designer_broad/Pred_ontgt_eff_allguides.svg',sep=''),width=5,height=2);
ggplot(subset(all_guides_df,!is.na(On.Target.Efficacy.Score)),
       aes(x=On.Target.Efficacy.Score))+
	geom_histogram(binwidth=0.02,fill='grey28')+
  theme_bw()+xlim(0,0.9)+
	xlab('Predicted activity score')+ylab('Number of guides')+
	facet_wrap(~gene)+
	theme(legend.position='none')
dev.off()


#Make a table of maximum target site conservation
max_cons<-group_by(all_guides_melted,subtype,gene)%>%
  summarize(max_hits=max(hits));
write.csv(max_cons,'./results_SpCas9/Max_conservation.csv',row.names=F);

#Make a table of total number of sequences used
alignseqs<-rbind(
  cbind(read.csv('./results_SpCas9_ltr/Aligned_seqs_subtypecounts_HIV1_ALL_2016_ltr_DNA.csv',
                 stringsAsFactors=F),data.frame(gene='ltr')),
  cbind(read.csv('./results_SpCas9_gag/Aligned_seqs_subtypecounts_HIV1_ALL_2016_gag_DNA.csv',
                 stringsAsFactors=F),data.frame(gene='gag')),
  cbind(read.csv('./results_SpCas9_pol/Aligned_seqs_subtypecounts_HIV1_ALL_2016_pol_DNA.csv',
                 stringsAsFactors=F),data.frame(gene='pol')));
seq_counts_summary<-melt(
  group_by(alignseqs,gene)%>%
  summarize(M=sum(tot_seqs),
          A=sum(tot_seqs[subtype%in%c('A','A1','A2','A3','A4','A5','A6')]),
          B=sum(tot_seqs[subtype=='B']),
          C=sum(tot_seqs[subtype=='C'])),
  id.vars='gene',variable.name='subtype',value.name='n_seqs');
write.csv(seq_counts_summary,'./alignments/Seq_counts_summary.csv',row.names=F);

temp<-merge(max_cons,seq_counts_summary)
temp$max_hits<-round(temp$max_hits,1)
temp$merge_col<-apply(temp,1,function(x)paste0(x['max_hits'],' (',x['n_seqs'],')'))
table_ppr<-reshape(temp[,c('subtype','gene','merge_col')],
                   idvar='gene',timevar='subtype',direction='wide');
write.csv(table_ppr,'./results_SpCas9/Max_conservation_combined.csv',row.names=F)
```



### 6. Multiplexed guides
Can run this after just  steps 1,2,4.
Guide pairs designed to target the max number of sequences overall (across grp M)
```{r}
library(dplyr);

for(grp in c('GpM','SubA','SubB','SubC','Other')){
	if(grp=='GpM'){
		match_matrix<-match_matrix_gpM;
	}else if(grp=='SubA'){
		match_matrix<-match_matrix_subA;
	}else if(grp=='SubB'){
		match_matrix<-match_matrix_subB;
	}else if(grp=='SubC'){
		match_matrix<-match_matrix_subC;
	}else if(grp=='Other'){
		match_matrix<-match_matrix_other;
	}
	suffix<-grp;
	thresh<-0.5*nrow(match_matrix); #threshold level of hits: pair should target at least half
	
	
	match_matrix_logical<-match_matrix>0;
	match_pairs_all<-apply(match_matrix_logical,2,function(x)
		apply(match_matrix_logical,2,function(y)sum(x|y)));
	match_pairs_all[lower.tri(match_pairs_all,diag=T)]<-0;
	match_pairs<-as.data.frame(which(match_pairs_all>thresh,arr.ind=T));
	match_pairs$pair_id<-apply(match_pairs,1,function(x){
		x<-as.list(x);return(paste(sort(c(x$row,x$col)),collapse='_'))});
	match_pairs<-distinct(match_pairs,pair_id,.keep_all=T);
	match_pairs$pair_freq<-match_pairs_all[match_pairs_all>thresh];
	match_pairs$guide1<-all_guides[match_pairs$row]; 
	match_pairs$guide2<-all_guides[match_pairs$col];
	match_pairs$g1_sing_freq<-all_guides_df$gpM_freq_total[match_pairs$row];
	match_pairs$g2_sing_freq<-all_guides_df$gpM_freq_total[match_pairs$col];
	
	#Check subtype specific hits for each pair
	match_pairs_all<-apply(match_matrix_subA>0,2,function(x)
		apply(match_matrix_subA>0,2,function(y)sum(x|y)));
	match_pairs$pair_freq_subA<-unlist(unname(apply(match_pairs,1,function(x){
		x<-as.list(x);match_pairs_all[as.numeric(x$row),as.numeric(x$col)]})));
	match_pairs_all<-apply(match_matrix_subB>0,2,function(x)
		apply(match_matrix_subB>0,2,function(y)sum(x|y)));
	match_pairs$pair_freq_subB<-unlist(unname(apply(match_pairs,1,function(x){
		x<-as.list(x);match_pairs_all[as.numeric(x$row),as.numeric(x$col)]})));
	match_pairs_all<-apply(match_matrix_subC>0,2,function(x)
		apply(match_matrix_subC>0,2,function(y)sum(x|y)));
	match_pairs$pair_freq_subC<-unlist(unname(apply(match_pairs,1,function(x){
		x<-as.list(x);match_pairs_all[as.numeric(x$row),as.numeric(x$col)]})));
	match_pairs_all<-apply(match_matrix_other>0,2,function(x)
		apply(match_matrix_other>0,2,function(y)sum(x|y)));
	match_pairs$pair_freq_other<-unlist(unname(apply(match_pairs,1,function(x){
		x<-as.list(x);match_pairs_all[as.numeric(x$row),as.numeric(x$col)]})));
	
	match_pairs<-mutate(match_pairs,
											pair_perc_GpM=100*pair_freq/nrow(match_matrix_gpM),
											pair_perc_SubA=100*pair_freq_subA/nrow(match_matrix_subA),
											pair_perc_SubB=100*pair_freq_subB/nrow(match_matrix_subB),
											pair_perc_SubC=100*pair_freq_subC/nrow(match_matrix_subC),
											pair_perc_other=100*pair_freq_other/nrow(match_matrix_other));
	
	match_pairs<-match_pairs[order(match_pairs$pair_freq,decreasing=T),];
	nrow(match_pairs)
	write.csv(match_pairs,file=paste(output_dir,'Guide_pairs_',suffix,'.csv',sep=''),row.names=F);
	
	#Top 20 guide pairs
	match_matrix_top<-t(vcountPDict(PDict(match_pairs$guide1[1:20]),reads[[1]]$seq,
																	max.mismatch=0,collapse=F)+
												vcountPDict(PDict(match_pairs$guide1[1:20]),reverseComplement(reads[[1]]$seq),
																		max.mismatch=0,collapse=F))+
		t(vcountPDict(PDict(match_pairs$guide2[1:20]),reads[[1]]$seq,
									max.mismatch=0,collapse=F)+
				vcountPDict(PDict(match_pairs$guide2[1:20]),reverseComplement(reads[[1]]$seq),
										max.mismatch=0,collapse=F));
	match_matrix_top[match_matrix_top>0]<-1;
	pdf(file=paste(output_dir,'Heatmap_guidepairs_top_',suffix,'.pdf',sep=''),width=10,height=3)
	par(mar=c(3,3,0.5,0.5))
	image(x=c(1:nrow(match_matrix_top)),y=c(1:ncol(match_matrix_top)),
				z=match_matrix_top,col=c('#f7f7f7','#7fbf7b'),
				xlab='',ylab='');
	rect(xleft=min(which(aligned_seqs_table$subtype%in%c('A1','A2','A'))),
			 ybottom=0,xright=max(which(aligned_seqs_table$subtype%in%c('A1','A2','A'))),
			 ytop=ncol(match_matrix_top)+0.5,col=rgb(255/255,182/255,193/255,0.3),border=NA)
	rect(xleft=min(which(aligned_seqs_table$subtype=='B')),
			 ybottom=0,xright=max(which(aligned_seqs_table$subtype=='B')),
			 ytop=ncol(match_matrix_top)+0.5,col=rgb(135/255,206/255,250/255,0.3),border=NA);
	rect(xleft=min(which(aligned_seqs_table$subtype=='C')),
			 ybottom=0,xright=max(which(aligned_seqs_table$subtype=='C')),
			 ytop=ncol(match_matrix_top)+0.5,col=rgb(255/255,255/255,0/255,0.3),border=NA);
	legend('topright',legend=c('A','B','C'),fill=c(rgb(255/255,182/255,193/255,0.3),
																								 rgb(135/255,206/255,250/255,0.3),
																								 rgb(255/255,255/255,0/255,0.3)),
				 bty='n',title='Subtype');
	mtext('LANL sequence',side=1,line=2);
	mtext('Guide pair',side=2,line=2);
	dev.off();
}

```


Triplets
```{r}
library(parallel);

for(grp in c('GpM','SubA','SubB','SubC','Other')){
	if(grp=='GpM'){
		match_matrix<-match_matrix_gpM;
	}else if(grp=='SubA'){
		match_matrix<-match_matrix_subA;
	}else if(grp=='SubB'){
		match_matrix<-match_matrix_subB;
	}else if(grp=='SubC'){
		match_matrix<-match_matrix_subC;
	}else if(grp=='Other'){
		match_matrix<-match_matrix_other;
	}
	suffix<-grp;
	
	triplet_thresh<-0.6*nrow(match_matrix); #threshold level of hits: triplets should target at least these many
	pair_thresh<-0.5*nrow(match_matrix); #pairs chosen should already target at least these many
	max_pairs<-1000; #maximum number of pairs to evaluate
	
	
	print(triplet_thresh);
	ncores=detectCores()
	# choose(length(all_guides),3); #To test all possible guide triplets would be nC3 combinations (about 3 million)
	
	#Instead, let's add each of the 265 guides to the top n guide pairs and see which ones improve the hits
	match_pairs<-read.csv(file=paste(output_dir,'Guide_pairs_',suffix,'.csv',sep=''),stringsAsFactors=F);
	# all_guides_df<-read.csv(file=paste(output_dir,'AllGuides.csv',sep=''),stringsAsFactors=F);
	all_guides<-all_guides_df$gseq;
	
	print(nrow(match_pairs));
	nrow(match_pairs)*(length(all_guides)-2);  #if we had to evaluate all pairs from previous step with all guides
	npairs<-min(sum(match_pairs$pair_freq>pair_thresh),max_pairs); #how many pairs to evaluate
	print(npairs);
	print(npairs*(length(all_guides)-2));
	nguides<-sum((pair_thresh+all_guides_df$gpM_freq_total)>triplet_thresh);#how many guides would improve triplet frequency to above threshold (best case)
	print(nguides) 
	print(npairs*nguides); #final number of combinations to test
	
	match_matrix_logical<-match_matrix>0;
	match_triplets_all<-mclapply(c(1:npairs),function(x)
		mclapply(c(1:nguides)[-c(match_pairs$row[x],match_pairs$col[x])],function(y)
			return(list(id=y,hits=sum(match_matrix_logical[,match_pairs$row[x]]|match_matrix_logical[,match_pairs$col[x]]|
																	match_matrix_logical[,y])))
			,mc.cores=ncores),mc.cores=ncores);
	
	#Only want triplets with hits frequency better than pairs or some other desired frequency
	min_hits<-triplet_thresh;
	tmp<-do.call('rbind',mclapply(c(1:length(match_triplets_all)),function(x){
		return(data.frame(g3_ind=unlist(lapply(match_triplets_all[[x]],function(z)return(z$id))),
											g1_ind=match_pairs$row[x],g2_ind=match_pairs$col[x],
											trip_freq=unlist(lapply(match_triplets_all[[x]],function(z)return(z$hits)))))
	},mc.cores=ncores));
	tmp<-tmp[tmp$trip_freq>=min_hits,];
	
	match_triplets<-data.frame(g1_ind=tmp$g1_ind,g2_ind=tmp$g2_ind,g3_ind=tmp$g3_ind,
														 guide1=all_guides[tmp$g1_ind],guide2=all_guides[tmp$g2_ind],
														 guide3=all_guides[tmp$g3_ind],
														 trip_freq=tmp$trip_freq,
														 trip_perc=100*tmp$trip_freq/length(reads[[1]]$seq),
														 g1_sing_freq=all_guides_df$gpM_freq_total[tmp$g1_ind],
														 g2_sing_freq=all_guides_df$gpM_freq_total[tmp$g2_ind],
														 g3_sing_freq=all_guides_df$gpM_freq_total[tmp$g3_ind],stringsAsFactors=F);
	nrow(match_triplets);
	
	#Remove duplicates
	match_triplets$triplet_id<-'';
	match_triplets$triplet_id<-apply(match_triplets,1,function(x){
		x<-as.list(x);return(paste(sort(as.numeric(c(x$g1_ind,x$g2_ind,x$g3_ind))),collapse='_'))});
	match_triplets<-distinct(match_triplets,triplet_id,.keep_all=T);
	match_triplets<-match_triplets[order(match_triplets$trip_freq,decreasing=T),]
	nrow(match_triplets)
	
	#Check subtype specific hits for each triplet
	match_matrix_logical<-match_matrix_subA>0;
	match_triplets$trip_freq_subA<-unlist(mclapply(c(1:nrow(match_triplets)),function(x)
		sum(match_matrix_logical[,match_triplets$g1_ind[x]]|
					match_matrix_logical[,match_triplets$g2_ind[x]]|
					match_matrix_logical[,match_triplets$g3_ind[x]]),mc.cores=ncores));
	match_matrix_logical<-match_matrix_subB>0;
	match_triplets$trip_freq_subB<-unlist(mclapply(c(1:nrow(match_triplets)),function(x)
		sum(match_matrix_logical[,match_triplets$g1_ind[x]]|
					match_matrix_logical[,match_triplets$g2_ind[x]]|
					match_matrix_logical[,match_triplets$g3_ind[x]]),mc.cores=ncores));
	match_matrix_logical<-match_matrix_subC>0;
	match_triplets$trip_freq_subC<-unlist(mclapply(c(1:nrow(match_triplets)),function(x)
		sum(match_matrix_logical[,match_triplets$g1_ind[x]]|
					match_matrix_logical[,match_triplets$g2_ind[x]]|
					match_matrix_logical[,match_triplets$g3_ind[x]]),mc.cores=ncores));
	match_matrix_logical<-match_matrix_other>0;
	match_triplets$trip_freq_other<-unlist(mclapply(c(1:nrow(match_triplets)),function(x)
		sum(match_matrix_logical[,match_triplets$g1_ind[x]]|
					match_matrix_logical[,match_triplets$g2_ind[x]]|
					match_matrix_logical[,match_triplets$g3_ind[x]]),mc.cores=ncores));
	
	match_triplets<-mutate(match_triplets,
												 trip_perc_GpM=100*trip_freq/nrow(match_matrix_gpM),
												 trip_perc_SubA=100*trip_freq_subA/nrow(match_matrix_subA),
												 trip_perc_SubB=100*trip_freq_subB/nrow(match_matrix_subB),
												 trip_perc_SubC=100*trip_freq_subC/nrow(match_matrix_subC),
												 trip_perc_other=100*trip_freq_other/nrow(match_matrix_other));
	
	nrow(match_triplets)
	write.csv(match_triplets[1:5000,],
						file=paste(output_dir,'Guide_triplets_',suffix,'.csv',sep=''),row.names=F);
	
	
	#Top 20 guide triples
	match_matrix_top<-
		t(vcountPDict(PDict(match_triplets$guide1[1:20]),reads[[1]]$seq,
									max.mismatch=0,collapse=F)+
				vcountPDict(PDict(match_triplets$guide1[1:20]),reverseComplement(reads[[1]]$seq),
										max.mismatch=0,collapse=F))+
		t(vcountPDict(PDict(match_triplets$guide2[1:20]),reads[[1]]$seq,
									max.mismatch=0,collapse=F)+
				vcountPDict(PDict(match_triplets$guide2[1:20]),reverseComplement(reads[[1]]$seq),
										max.mismatch=0,collapse=F))+
		t(vcountPDict(PDict(match_triplets$guide3[1:20]),reads[[1]]$seq,
									max.mismatch=0,collapse=F)+
				vcountPDict(PDict(match_triplets$guide3[1:20]),reverseComplement(reads[[1]]$seq),
										max.mismatch=0,collapse=F));
	match_matrix_top[match_matrix_top>0]<-1;
	pdf(file=paste(output_dir,'Heatmap_guidetriples_top_',suffix,'.pdf',sep=''),width=10,height=3)
	par(mar=c(3,3,0.5,0.5))
	image(x=c(1:nrow(match_matrix_top)),y=c(1:ncol(match_matrix_top)),
				z=match_matrix_top,col=c('#f7f7f7','#7fbf7b'),
				xlab='',ylab='');
	rect(xleft=min(which(aligned_seqs_table$subtype%in%c('A1','A2','A'))),
			 ybottom=0,xright=max(which(aligned_seqs_table$subtype%in%c('A1','A2','A'))),
			 ytop=ncol(match_matrix_top)+0.5,col=rgb(255/255,182/255,193/255,0.3),border=NA)
	rect(xleft=min(which(aligned_seqs_table$subtype=='B')),
			 ybottom=0,xright=max(which(aligned_seqs_table$subtype=='B')),
			 ytop=ncol(match_matrix_top)+0.5,col=rgb(135/255,206/255,250/255,0.3),border=NA);
	rect(xleft=min(which(aligned_seqs_table$subtype=='C')),
			 ybottom=0,xright=max(which(aligned_seqs_table$subtype=='C')),
			 ytop=ncol(match_matrix_top)+0.5,col=rgb(255/255,255/255,0/255,0.3),border=NA);
	legend('topright',legend=c('A','B','C'),fill=c(rgb(255/255,182/255,193/255,0.3),
																								 rgb(135/255,206/255,250/255,0.3),
																								 rgb(255/255,255/255,0/255,0.3)),
				 bty='n',title='Subtype');
	mtext('LANL sequence',side=1,line=2);
	mtext('Guide triplet',side=2,line=2);
	dev.off();
	
}
```


For paper, maximum % sequences targeted by pairs and triplets across genes and subtypes
This chunk can be run independently
```{r}
require(parallel); ncores<-detectCores()
pairs<-do.call('rbind',mclapply(c('ltr','gag','pol'),function(gene)
	do.call('rbind',mclapply(c('GpM','SubA','SubB','SubC','Other'),function(grp)
		cbind(read.csv(paste0('./results_SpCas9_',gene,'/Guide_pairs_',grp,'.csv'),stringsAsFactors=F)[1:5,],data.frame(group=grp,gene=gene)),
		mc.cores=ncores)),mc.cores=ncores));
head(pairs);
dim(pairs);
write.csv(pairs,'./results_SpCas9/Top5_pairs_all.csv',row.names=F)


pairs_summary<-
	do.call('rbind',lapply(c('ltr','gag','pol'),function(gene)
	return(data.frame(gene=gene,
										max_M=max(pairs[pairs$group=='GpM'&pairs$gene==gene,'pair_perc_GpM']),
										max_A=max(pairs[pairs$group=='SubA'&pairs$gene==gene,'pair_perc_SubA']),
										max_B=max(pairs[pairs$group=='SubB'&pairs$gene==gene,'pair_perc_SubB']),
										max_C=max(pairs[pairs$group=='SubC'&pairs$gene==gene,'pair_perc_SubC']),
										max_other=max(pairs[pairs$group=='Other'&pairs$gene==gene,'pair_perc_other'])))));
write.csv(pairs_summary,'./results_SpCas9/Pairs_summary_all.csv',row.names=F)

triplets<-do.call('rbind',mclapply(c('ltr','gag','pol'),function(gene)
	do.call('rbind',mclapply(c('GpM','SubA','SubB','SubC','Other'),function(grp)
		cbind(read.csv(paste0('./results_SpCas9_',gene,'/Guide_triplets_',grp,'.csv'),stringsAsFactors=F)[1:5,],data.frame(group=grp,gene=gene)),
		mc.cores=ncores)),mc.cores=ncores));
head(triplets);
dim(triplets);
write.csv(triplets,'./results_SpCas9/Top5_triplets_all.csv',row.names=F)


triplets_summary<-
	do.call('rbind',lapply(c('ltr','gag','pol'),function(gene)
	return(data.frame(gene=gene,
										max_M=max(triplets[triplets$group=='GpM'&triplets$gene==gene,'trip_perc_GpM']),
										max_A=max(triplets[triplets$group=='SubA'&triplets$gene==gene,'trip_perc_SubA']),
										max_B=max(triplets[triplets$group=='SubB'&triplets$gene==gene,'trip_perc_SubB']),
										max_C=max(triplets[triplets$group=='SubC'&triplets$gene==gene,'trip_perc_SubC']),
										max_other=max(triplets[triplets$group=='Other'&triplets$gene==gene,'trip_perc_other'])))));
write.csv(triplets_summary,'./results_SpCas9/Triplets_summary_all.csv',row.names=F)

```




### 8. Now make a list of guides to test
```{r}
inds<-unique(c(1:20,
               order(all_guides_df$subA_freq_total,decreasing=T)[1:10],
               order(all_guides_df$subB_freq_total,decreasing=T)[1:10],
               order(all_guides_df$subC_freq_total,decreasing=T)[1:10],
               order(all_guides_df$other_freq_total,decreasing=T)[1:10],
               unique(c(sapply(c('GpM','SubA','SubB','SubC','Other'),function(x)
                 unname(c(read.csv(paste(output_dir,'Guide_pairs_',x,'.csv',sep=''),stringsAsFactors=F)
                        [1:20,c('row','col')],recursive=T))),recursive=T)),
               unique(c(sapply(c('GpM','SubA','SubB','SubC','Other'),function(x)
                 unname(c(read.csv(paste(output_dir,'Guide_triplets_',x,'.csv',sep=''),stringsAsFactors=F)
                        [1:20,c('g1_ind','g2_ind','g3_ind')],recursive=T))),recursive=T))));
test_guides<-all_guides_df[inds,]%>%select(gid,guide,pam,dir,from_consensus,gpM_perc:subC_perc,
                                         benchling_efficiency:off_5);
nrow(test_guides)
write.csv(test_guides,file=paste(output_dir,'GuidesToTest.csv',sep=''),row.names=F);
```






